client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java

```diff
@@ -240,7 +277,7 @@ private void releaseSegmentsIfNeeded(PositionInternal position) throws ReaderNot
             if (reader != null) {
                 if (groupState.releaseSegment(segment, reader.getOffset(), getLag(), position)) {
                     readers.remove(reader);
-                    ranges.remove(reader.getSegmentId());
+                    ranges = CopyOnWriteMapUtils.remove(ranges, reader.getSegmentId());
```

Comment:
Given that `ranges` is guarded by readers   `@GuardedBy("readers")` why do we need to create a copy? isn't it redundant ?

Reply:
The reason is that the `ranges` map is referenced by many `PositionImpl` objects (i.e, every time that `readNextEvent()` returns an event and there is no change in segments, the same `ranges` map is reused). This means that we cannot modify this map if there is a change in segments, as it would impact to all the previous `PositionImpl` objects and they would be in an inconsistent state. So, every time there is a change in the segments, we need the variable `ranges` to point to a new map object and apply the change in segments to the new map.